lib/githubbot.ts

```diff
@@ -0,0 +1,403 @@
+                    login: string;
+                }
+            },
+            sha: string
+        }
+    }
+};
+
+
+// Github API -----------------------------------------------------------------
+// These provide the current bare minimum definitions for child Procbots working with them.
+export interface CommitFile {
+    filename: string,
+};
+
+export interface Commit {
+    commit: {
+        committer: {
+            name: string
+        }
+        message: string
+    },
+    files: CommitFile[],
+    sha: string
+};
+
+export interface Review {
+    state: string
+};
+
+export interface Merge {
+    sha: string
+};
+
+export interface Tag {
+    sha: string
+};
+
+export interface PullRequest {
+    head: {
+        ref: string
+    }
+};
+
+export interface Blob {
+    sha: string
+};
+
+export interface TreeEntry {
+    path: string,
+    mode: string,
+    type: string,
+    sha: string,
+    url?: string,
+    size?: number
+};
+
+export interface Tree {
+    sha: string,
+    url: string,
+    tree: TreeEntry[]
+};
+
+// GithubBot ------------------------------------------------------------------
+
+// The GithubAction  defines an action, which is passed to a WorkerMethod should all of
+// the given pre-requisites be applicable.
+// At least one event. For each additional event, this is considered an 'OR' comparator.
+// eg. 'pull_request' OR 'pull_request_review' event
+// Labels, on the other hand, are ANDed together.
+// eg. 'flow/ready-to-merge' AND 'flow/in-review' (trigger or suppression)
+//
+// In the future, some sort of comparator language might be useful, eg:
+// and: [
+//  {
+//      or: [
+//          {
+//              name: <eventType>
+//              value: 'pull_request',
+//              op: eq | neq;
+//          },
+//          ...
+//      }]
+//  ,
+//  ...
+//  }
+// ]
+export interface GithubAction {
+    name: string,
+    events: string[],
+    triggerLabels?: string[],
+    suppressionLabels?: string[]
+}
+
+// The Register interface is passed to the GithubBot.register method to register
+// for callback when the appropriate events and labels are received.
+export interface GithubActionRegister extends GithubAction {
+    workerMethod: GithubActionMethod
+}
+
+// A GithubActionMethod is the method that will be used to process an event.
+export type GithubActionMethod = <T>(action: GithubAction, data: T) => Promise<void>;
+
+// Main GithubBot.
+export class GithubBot extends ProcBot.ProcBot<string> {
+    private integrationId: number;
+    private user: number;
+    private eventTriggers: GithubActionRegister[] = [];
+    protected githubApi: any;
+
+    // Takes a set of webhook types that the bot is interested in.
+    // Registrations can be passed in on bot creation, or registered/deregistered later.
+    // However, 'baked in' registrations are not available for deregistration.
+    constructor(integration: number) {
+        super();
+        this._botname = 'GithubBot';
```

Comment:
The setting of `_botname` feels like a strangely secret interface of ProcBot. Why not make it an optional constructor argument, so you can just pass it to `super()`?